### PR TITLE
updated lm1b example with jit

### PR DESCRIPTION
**Pull request by chiamp**
Updated lm1b example to use jit instead of pmap.
lm1b now uses both data and model parallelism instead of just data parallelism.

### CODE REVIEW

File: examples/lm1b/configs/default.py [hunk 1]
```diff
@@ -115,6 +115,35 @@ def get_config():
   config.seed = 0
 
   # Prompt for language model sampling.
-  config.prompts = "I love to "
+  config.prompts = 'I love to '
+
+  # Parallelism
```
Comment (marcvanzee) [inline examples/lm1b/configs/default.py:120]: Maybe mention you got these for maxtext and link to it.

File: examples/lm1b/configs/default.py [hunk 2]
```diff
@@ -115,6 +115,35 @@ def get_config():
...
+      ['activation_mlp', 'tensor'],
+      ['activation_heads', 'tensor'],
+      ['activation_kv', 'tensor'],
+      ['activation_vocab', 'tensor'],
+      ['mlp', 'tensor'],
+      ['vocab', 'tensor'],
+      ['embed', 'fsdp'],
+      ['heads', 'tensor'],
+  ]
+  config.full_sharding = [['data', 'fsdp', 'tensor']]
```
Comment (marcvanzee) [inline examples/lm1b/configs/default.py:135]: Why is this a list of lists and later used with * expansion? Why not simple a list and pass it to `NamedSharding` without expanding it? Same for `data_sharding` below.

File: examples/lm1b/configs/default.py [hunk 3]
```diff
@@ -115,6 +115,35 @@ def get_config():
...
+  config.data_sharding = [['data']]
+
+  # One axis for each parallelism type may hold a placeholder (-1)
+  # value to auto-shard based on available slices and devices.
+  # By default, product of the DCN axes should equal number of slices
```
Comment (marcvanzee) [inline examples/lm1b/configs/default.py:140]: Maybe explain what DCN and ICI are, I suppose not everyone knows this.

File: examples/lm1b/train.py [hunk 4]
```diff
@@ -527,10 +556,11 @@ def encode_strings(strs, max_len):
 
       # Shard data to devices and do a training step.
       with jax.profiler.StepTraceAnnotation("train", step_num=step):
-        batch = common_utils.shard(
-            jax.tree_util.tree_map(np.asarray, next(train_iter))
+        batch = next(train_iter)
+        batch = jax.tree_map(lambda x: jnp.array(x), batch)
```
Comment (marcvanzee) [inline examples/lm1b/train.py:560]: here you cast to jax array where the original seems to use numpy arrays, why did you do this?

File: examples/lm1b/train.py [hunk 5]
```diff
@@ -527,10 +556,11 @@ def encode_strings(strs, max_len):
...
+        state, metrics = p_train_step(
+            state, batch, train_config, learning_rate_fn, 0.0, dropout_rngs
```
Comment (marcvanzee) [inline examples/lm1b/train.py:562]: why do you pass 0.0 here? isn't this the default value? also I would always use kwarg here for a number.

File: examples/lm1b/train.py [hunk 6]
```diff
@@ -542,7 +572,7 @@ def encode_strings(strs, max_len):
       if step % config.eval_every_steps == 0 or is_last_step:
         with report_progress.timed("training_metrics"):
           logging.info("Gathering training metrics.")
-          train_metrics = common_utils.get_metrics(train_metrics)
+          train_metrics = common_utils.stack_forest(train_metrics)
```
Comment (marcvanzee) [inline examples/lm1b/train.py:575]: why was this necessary?

File: examples/lm1b/utils.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+import jax.numpy as jnp
+from jax.experimental import mesh_utils
+
+
+# Mesh utils.
+# -----------------------------------------------------------------------------
+
+
+def create_device_mesh(config):
+  """Creates a device mesh with each slice in its own data parallel group. If there is only one slice, uses two replicas"""
```
Comment (marcvanzee) [inline examples/lm1b/utils.py:18]:
```suggestion
  """Creates a device mesh with each slice in its own data parallel group. If there is only one slice, uses two replicas."""
```

File: examples/lm1b/utils.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
+import functools
```
Comment (marcvanzee) [inline examples/lm1b/utils.py:1]: Is this file copied from maxtext? Please add a comment if so.

**Review by marcvanzee — CHANGES_REQUESTED** [review]

File: examples/lm1b/train.py (hunk 4 above)
Comment (chiamp) [reply examples/lm1b/train.py:562]: I get this error without casting:
```
File "/home/marcuschiam/flax/examples/lm1b/train.py", line 565, in train_and_evaluate
    state, metrics = jit_train_step(
TypeError: dtype tf.int32 not understood
```
The original element type is `EagerTensor` I think, with dtype `tf.int32`.

File: examples/lm1b/train.py (hunk 5 above)
Comment (chiamp) [reply examples/lm1b/train.py:564]: If I try:
```
state, metrics = jit_train_step(
            state, batch, train_config, learning_rate_fn, dropout_rng=dropout_rngs
        )
```
I get:
```
File "/home/marcuschiam/flax/examples/lm1b/train.py", line 565, in train_and_evaluate
    state, metrics = jit_train_step(
ValueError: pjit does not support kwargs when in_shardings is specified.
```

File: examples/lm1b/train.py (hunk 6 above)
Comment (chiamp) [reply examples/lm1b/train.py:577]: [`get_metrics`](https://github.com/google/flax/blob/main/flax/training/common_utils.py#L72) is a utility function for pmap to get a single copy of a device-replicated metric. It calls `stack_forest` on the metrics.
If I try using `get_metrics` with `jit`, I get:
```
File "/home/marcuschiam/.local/lib/python3.8/site-packages/flax/training/common_utils.py", line 83, in <lambda>
    device_metrics = jax.tree_util.tree_map(lambda x: x[0], device_metrics)
  File "/home/marcuschiam/.local/lib/python3.8/site-packages/jax/_src/array.py", line 317, in __getitem__
    return lax_numpy._rewriting_take(self, idx)
  File "/home/marcuschiam/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4143, in _rewriting_take
    return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
  File "/home/marcuschiam/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4152, in _gather
    indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
  File "/home/marcuschiam/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4253, in _index_to_gather
    idx = _canonicalize_tuple_index(len(x_shape), idx)
  File "/home/marcuschiam/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4586, in _canonicalize_tuple_index
    raise IndexError(
IndexError: Too many indices for array: 1 non-None/Ellipsis indices for dim 0.
```

File: examples/lm1b/train.py (hunk 4 above)
Comment (marcvanzee) [reply examples/lm1b/train.py:562]: Ok makes sense!

File: examples/lm1b/train.py (hunk 6 above)
Comment (marcvanzee) [reply examples/lm1b/train.py:577]: Cool, thanks!

File: examples/lm1b/train.py (hunk 5 above)
Comment (marcvanzee) [reply examples/lm1b/train.py:564]: Hmm ok, thanks for letting me know.

**Review by marcvanzee — APPROVED** [review]
